### PR TITLE
Support for lazy modules

### PR DESCRIPTION
**Pull request by mlazos**
Added fixes to support [lazy modules](https://pytorch.org/docs/stable/generated/torch.nn.modules.lazy.LazyModuleMixin.html) in torchdynamo. The main issue that needed to be addressed is that LazyModules register some hooks which are run when a module is called. Torchdynamo typically calls the `forward` method instead of `__call__`  so these hooks were never run. In the case of LazyModules we now run and trace the `__call__` method, and allow the original module to be mutated. In the future, we could do this for all modules, but there were cases where torchdynamo does not yet support functionality used in all hooks.

### CODE REVIEW

File: tests/test_modules.py [hunk 1]
```diff
@@ -776,3 +805,36 @@ def forward(self, x):
...
+    def test_lazy_module(self):
+        input_shape = (16, 3, 6, 7, 8)
+        cnt = torchdynamo.testing.CompileCounter()
+        with torchdynamo.optimize(cnt):
+
+            def test_dynamic_module():
+                module = LazyModule()
+                input = torch.ones(*input_shape)
+                module(input)
+                self.assertTrue(
```
Comment (jansel) [inline tests/test_modules.py:818]: Let's move the assert outside torchdynamo.optimize()

File: tests/test_modules.py [hunk 2]
```diff
@@ -776,3 +805,36 @@ def forward(self, x):
...
+
+    def test_lazy_module(self):
+        input_shape = (16, 3, 6, 7, 8)
+        cnt = torchdynamo.testing.CompileCounter()
+        with torchdynamo.optimize(cnt):
+
+            def test_dynamic_module():
+                module = LazyModule()
+                input = torch.ones(*input_shape)
+                module(input)
```
Comment (jansel) [inline tests/test_modules.py:817]: we should call this multiple times and assert cnt.frame_count is 1 (to check for guards failing)

File: tests/test_modules.py [hunk 3]
```diff
@@ -776,3 +805,36 @@ def forward(self, x):
...
         self.assertTrue(torchdynamo.testing.same(pre, opt_pre))
         self.assertTrue(torchdynamo.testing.same(out1, out_post))
+
+    def test_lazy_module(self):
+        input_shape = (16, 3, 6, 7, 8)
+        cnt = torchdynamo.testing.CompileCounter()
+        with torchdynamo.optimize(cnt):
+
+            def test_dynamic_module():
+                module = LazyModule()
```
Comment (jansel) [inline tests/test_modules.py:815]: Construct this outside of torchdynamo.optimize(), otherwise it gets mapped to UnspecializedNNModule()

**Review by jansel — CHANGES_REQUESTED** [review]

File: tests/test_modules.py (hunk 1 above)
Comment (jansel) [reply tests/test_modules.py:818]: To make sure it isn't getting optimized away.

File: tests/test_modules.py (hunk 3 above)
Comment (mlazos) [reply tests/test_modules.py:815]: Do I want to test that case too? I also have a check in UnspecializedNNModule to handle lazy modules.

Comment (jansel) [reply tests/test_modules.py:834]: Yes we should check them both.

Comment (mlazos) [reply tests/test_modules.py:834]: Should be good now

**Pull request merged**